Add QA with Langchain of AnalyticDB  vector store and OpenAI example.

wangxuqi opened the pull request:
Hi there,I hope you're doing well. 

I'm excited to open this PR. This is another example of using a fully Postgres syntax compatible database 'AnalyticDB' as a datastore to build QA with Langchain and OpenAI APIs. 

As AnalyticDB can be used with Langchain directly for now. This example is a better case for user who want to use  'AnalyticDB' as a vector store.

I would really appreciate your feedback.

ted-at-openai commented:
Thanks, I'll take a look. What's your goal for this contribution?

ted-at-openai commented:
Also, what kind of feedback are you looking for? If you think it's in good shape, I'll spend less time reviewing it before merging.

wangxuqi commented:
> Also, what kind of feedback are you looking for? If you think it's in good shape, I'll spend less time reviewing it before merging.

So sorry if my commit message wasn't clear enough in conveying my intentions.  I think it's in good shape, This notebook presents how to implement a Question Answering system with LangChain, AnalyticDB and OpenAI embeddings. This example mainly consists of the following main steps.
- prepare connection string with Environment variables and LangChain
```
from langchain.vectorstores.analyticdb import AnalyticDB

CONNECTION_STRING = AnalyticDB.connection_string_from_db_params(
    driver=os.environ.get("PG_DRIVER", "psycopg2cffi"),
    host=os.environ.get("PG_HOST", "localhost"),
    port=int(os.environ.get("PG_PORT", "5432")),
    database=os.environ.get("PG_DATABASE", "postgres"),
    user=os.environ.get("PG_USER", "postgres"),
    password=os.environ.get("PG_PASSWORD", "postgres"),
)
```
- Definition a VectorStore Chain of AnalyticDB
```
from langchain.vectorstores import AnalyticDB
from langchain.embeddings import OpenAIEmbeddings
from langchain import VectorDBQA, OpenAI

embeddings = OpenAIEmbeddings()
doc_store = AnalyticDB.from_texts(
    texts=answers, embedding=embeddings, connection_string=CONNECTION_STRING,
    pre_delete_collection=True,
)
```
- Define QA chain with OpenAI llm and vector store defined above.
```
llm = OpenAI()
qa = VectorDBQA.from_chain_type(
    llm=llm,
    chain_type="stuff",
    vectorstore=doc_store,
    return_source_documents=False,
)
```
- SearchData with the QA chain
```
qa.run(question)
```

ted-at-openai commented on examples/vector_databases/analyticdb/QA_with_Langchain_AnalyticDB_and_OpenAI.ipynb:
```diff
@@ -0,0 +1,442 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Question Answering with Langchain, AnalyticDB and OpenAI\n",
```
Want to make this a markdown header?

ted-at-openai commented on examples/vector_databases/analyticdb/QA_with_Langchain_AnalyticDB_and_OpenAI.ipynb:
```diff
@@ -0,0 +1,442 @@
+   "metadata": {},
+   "source": [
+    "Question Answering with Langchain, AnalyticDB and OpenAI\n",
+    "This notebook presents how to implement a Question Answering system with Langchain, AnalyticDB as a knowledge based and OpenAI embeddings. If you are not familiar with AnalyticDB, it’s better to check out the Getting_started_with_AnalyticDB_and_OpenAI.ipynb notebook.\n",
```
Optional: can add a markdown link to the other notebook

ted-at-openai commented on examples/vector_databases/analyticdb/QA_with_Langchain_AnalyticDB_and_OpenAI.ipynb:
```diff
@@ -0,0 +1,442 @@
+   "source": [
+    "Prerequisites\n",
+    "For the purposes of this exercise we need to prepare a couple of things:\n",
+    "[AnalyticDB cloud instance](https://www.alibabacloud.com/help/en/analyticdb-for-postgresql/latest/product-introduction-overview).\n",
```
Recommend making this a markdown bulleted list

ted-at-openai commented on examples/vector_databases/analyticdb/QA_with_Langchain_AnalyticDB_and_OpenAI.ipynb:
```diff
@@ -0,0 +1,442 @@
+   "metadata": {},
+   "source": [
+    "### Install requirements\n",
+    "This notebook obviously requires the `openai`, `langchain`, `psycopg2cffi` and `tiktoken`"
```
optional: explain what each of these is for, add links (I wasn't sure what psycopg2ffi was for and had to look it up)

ted-at-openai commented on examples/vector_databases/analyticdb/QA_with_Langchain_AnalyticDB_and_OpenAI.ipynb:
```diff
@@ -0,0 +1,442 @@
+    "### Prepare your OpenAI API key\n",
+    "The OpenAI API key is used for vectorization of the documents and queries.\n",
+    "\n",
+    "If you don't have an OpenAI API key, you can get one from [https://beta.openai.com/account/api-keys](https://beta.openai.com/account/api-keys).\n",
```
optional: new url is https://platform.openai.com/account/api-keys (platform instead of beta)

ted-at-openai commented on examples/vector_databases/analyticdb/QA_with_Langchain_AnalyticDB_and_OpenAI.ipynb:
```diff
@@ -0,0 +1,442 @@
+   "metadata": {},
+   "source": [
+    "### Prepare your AnalyticDB connection string\n",
+    "The AnalyticDB connection string is build with PG_HOST, PG_PORT, PG_DATABASE, PG_USER, PG_PASSWORD. you need to export them first to set correct connect string.And then build the connection string"
```
"You" missing capitalization. ".And" missing space.

ted-at-openai replied:
final word missing period.

ted-at-openai commented on examples/vector_databases/analyticdb/QA_with_Langchain_AnalyticDB_and_OpenAI.ipynb:
```diff
@@ -0,0 +1,442 @@
+   ],
+   "source": [
+    "! export PG_HOST=\"your AnalyticDB host url\"\n",
+    "! export PG_PORT=5432 #Optional defalut value is 5432\n",
```
"default" misspelled

ted-at-openai commented on examples/vector_databases/analyticdb/QA_with_Langchain_AnalyticDB_and_OpenAI.ipynb:
```diff
@@ -0,0 +1,442 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Load data\n",
```
Can be markdown header

ted-at-openai reviewed (changes requested):
Looks great. I didn't run the whole thing myself, as I didn't want to go through the trouble of setting up a DB. I left some suggestions mostly on wording and formatting. Happy to merge once addressed. Thank you!

wangxuqi commented:
> Looks great. I didn't run the whole thing myself, as I didn't want to go through the trouble of setting up a DB. I left some suggestions mostly on wording and formatting. Happy to merge once addressed. Thank you!
Thank you for your feedback! I've addressed your suggestions regarding wording and formatting. Thank you once again for your guidance and support, and I look forward to our continued collaboration.

wangxuqi commented:
I also have a related PR in another project : https://github.com/openai/chatgpt-retrieval-plugin/pull/147. I would appreciate it if you could take some time to review it.

ted-at-openai reviewed (approved):
Didn't check carefully, but I trust you addressed the changes and are happy with the final result. Will merge.
